File: leetcode/array/153.Find-Minimum-in-Rotated-Sorted-Array/153.Find-Minimum-in-Rotated-Sorted-Array.go
package leetcode

/* 153. 寻找旋转排序数组中的最小值
已知一个长度为 n 的数组，预先按照升序排列，经由 1 到 n 次 旋转 后，
得到输入数组。例如，原数组 nums = [0,1,2,4,5,6,7] 在变化后可能得到：
若旋转 4 次，则可以得到 [4,5,6,7,0,1,2]
若旋转 7 次，则可以得到 [0,1,2,4,5,6,7]
注意，数组 [a[0], a[1], a[2], ..., a[n-1]] 旋转一次 的结果为数组 [a[n-1], a[0], a[1], a[2], ..., a[n-2]] 。

给你一个元素值 互不相同 的数组 nums ，它原来是一个升序排列的数组，
并按上述情形进行了多次旋转。请你找出并返回数组中的 最小元素 。
你必须设计一个时间复杂度为 O(log n) 的算法解决此问题。

示例 1：
输入：nums = [3,4,5,1,2]
输出：1
解释：原数组为 [1,2,3,4,5] ，旋转 3 次得到输入数组。

示例 2：
输入：nums = [4,5,6,7,0,1,2]
输出：0
解释：原数组为 [0,1,2,4,5,6,7] ，旋转 4 次得到输入数组。

示例 3：
输入：nums = [11,13,15,17]
输出：11
解释：原数组为 [11,13,15,17] ，旋转 4 次得到输入数组。
*/

// 解法1 二分 不太好理解
func findMin(nums []int) int {
	low, high := 0, len(nums)-1
	for low < high {
		if nums[low] < nums[high] { // 此时数组为升序，输出第一个元素即可
			return nums[low]
		}
		mid := low + ((high - low) >> 1)
		if nums[mid] >= nums[low] { // 说明mid处的值还没有达到最大，mid在最小值的左边，向右缩小左边界
			low = mid + 1
		} else { // 说明mid在最小值的右边，此刻向左收缩右边界
			high = mid
		}
	}
	return nums[low]
}

// 解法2 暴力 时间复杂度 O(n)
func findMin2(nums []int) int {
	min := nums[0]
	for _, num := range nums[1:] {
		if min > num {
			min = num
		}
	}
	return min
}

// 解法3 便于理解的二分法
func findMin1(nums []int) int {
	if len(nums) == 0 { // 边界条件：nums无元素
		return 0
	}
	if len(nums) == 1 { // 边界条件：nums仅有一个元素
		return nums[0]
	}
	if nums[len(nums)-1] > nums[0] { // 由于事先排序了，如果最后一个元素大于第一个元素，说明没有发生旋转
		return nums[0]
	}

	low, high := 0, len(nums)-1
	for low <= high {
		mid := low + ((high - low) >> 1)
		if nums[low] < nums[high] {
			return nums[low]
		}
		if (mid == len(nums)-1 && nums[mid-1] > nums[mid] || (mid < len(nums)-1) && mid > 0 && nums[mid-1] > nums[mid] && nums[mid] < nums[mid+1]) {
			return nums[mid]
		}
		if nums[mid] > nums[low] && nums[low] > nums[high] { // mid 在数值大的一部分区间里
			low = mid + 1
		} else if nums[mid] < nums[low] && nums[low] > nums[high] { // mid 在数值小的一部分区间里
			high = mid - 1
		} else {
			if nums[low] == nums[mid] {
				low++
			}
			if nums[high] == nums[mid] {
				high--
			}
		}
	}
	return -1
}

/*
解题思路

数组旋转1次，相当于将数组最后一个元素挪到第一个位置

给出一个原本从小到大排序过的数组，但是在某一个分割点上，
把数组切分后的两部分对调位置，数值偏大的放到了数组的前部。
求这个数组中最小的元素。
求数组最小的元素其实就是找分割点，前一个数比当前数大，后一个数比当前数也要大。
可以用二分搜索查找，需要查找的两个有序区间。时间复杂度 O(log n)。
这一题也可以用暴力解法，从头开始遍历，动态维护一个最小值即可，时间复杂度 O(n)。
*/

File: leetcode-cpp/153.Find-Minimum-in-Rotated-Sorted-Array/153.Find-Minimum-in-Rotated-Sorted-Array.cpp
/**
 * @file 153.Find-Minimum-in-Rotated-Sorted-Array.cpp
 * @author your name (you@domain.com)
 * @brief 
 * @version 0.1
 * @date 2022-08-12
 * 
 * @copyright Copyright (c) 2022
 * leetcode-153. 寻找旋转排序数组中的最小值
 * 

 */

class Solution {
public:
    int findMin(vector<int>& nums) {
        int n = nums.size()-1;
        // if (n == 0) {
        //     return nums[0];
        // }
        // int res;
        int left = 0, right = n;
        while (left < right) {
            int mid = left + ((right - left) >> 1);
            if (nums[mid] < nums[right]) { // 动态的右边界
                right = mid;
            } else { // nums无相同的元素
                left = mid + 1;
            }
        }
        return nums[left];
    }
};
/**
 * @brief 
 * 方法一：二分查找
思路与算法
一个不包含重复元素的升序数组在经过旋转之后，
可以得到下面可视化的折线图：

其中横轴表示数组元素的下标，纵轴表示数组元素的值。
图中标出了最小值的位置，是我们需要查找的目标。

我们考虑数组中的最后一个元素 x：
在最小值右侧的元素（不包括最后一个元素本身），
它们的值一定都严格小于 x；而在最小值左侧的元素，它们的值一定都严格大于 x。
因此，我们可以根据这一条性质，通过二分查找的方法找出最小值。

在二分查找的每一步中，左边界为 low，右边界为 high，
区间的中点为 pivot，最小值就在该区间内。
我们将中轴元素 nums[pivot] 与右边界元素 nums[high] 进行比较，
可能会有以下的三种情况：

- 第一种情况是 nums[pivot]<nums[high]。
如下图所示，这说明 nums[pivot] 是最小值右侧的元素，
因此我们可以忽略二分查找区间的右半部分。

- 第二种情况是 nums[pivot]>nums[high]。
如下图所示，这说明 nums[pivot] 是最小值左侧的元素，
因此我们可以忽略二分查找区间的左半部分。

由于数组不包含重复元素，并且只要当前的区间长度不为 1，
pivot 就不会与 high 重合；而如果当前的区间长度为 1，
这说明我们已经可以结束二分查找了。
因此不会存在 nums[pivot]=nums[high] 的情况。

当二分查找结束时，我们就得到了最小值所在的位置。

复杂度分析
时间复杂度：时间复杂度为 O(logn)，其中 n 是数组 nums 的长度。
在二分查找的过程中，每一步会忽略一半的区间，
因此时间复杂度为 O(logn)。

空间复杂度：O(1)。
 */